fastchat/serve/openai_api_server.py

```diff
@@ -244,47 +385,130 @@ async def chat_completion(model_name: str, gen_params: Dict[str, Any]):
-    payload = {
-        "model": request.model,
-        "prompt": request.prompt,
-        "temperature": request.temperature,
-        "max_tokens": request.max_tokens,
-        "logprobs": request.logprobs,
-    }
+    error_check_ret = await check_model(request)
+    if error_check_ret is not None:
+        return error_check_ret
+    error_check_ret = check_requests(request)
+    if error_check_ret is not None:
+        return error_check_ret
+
+    payload = get_gen_params(
+        request.model,
+        request.prompt,
+        temperature=request.temperature,
+        top_p=request.top_p,
+        max_tokens=request.max_tokens,
+        echo=request.echo,
+        stream=request.stream,
+        stop=request.stop,
+    )
 
     if request.stream:
-        raise NotImplementedError("streaming is not supported yet")
+        generator = generate_completion_stream_generator(payload, request.n)
+        return StreamingResponse(generator, media_type="text/event-stream")
     else:
-        completions = []
-        prompt_tokens = 0
-        completion_tokens = 0
+        text_completions = []
         for i in range(request.n):
-            content = await generate_completion(payload)
-            content = json.loads(content)
-            content["index"] = i
-            completion_tokens += content["completion_tokens"]
-            prompt_tokens = content["prompt_tokens"]
-            content.pop("completion_tokens")
-            content.pop("prompt_tokens")
-            if request.echo:
-                content["text"] = request.prompt + content["text"]
-            completions.append(content)
-    return CompletionResponse(
-        model=request.model,
-        choices=completions,
-        usage={
-            "prompt_tokens": prompt_tokens,
-            "completion_tokens": completion_tokens,
-            "total_tokens": prompt_tokens + completion_tokens,
-        },
-    )
+            content = asyncio.create_task(generate_completion(payload))
+            text_completions.append(content)
+
+        choices = []
+        usage = {
+            "prompt_tokens": 0,
+            "completion_tokens": 0,
+            "total_tokens": 0,
+        }
+        for i, content_task in enumerate(text_completions):
+            try:
+                content = await content_task
+            except Exception as e:
+                return create_error_response(
+                    50090,
+                    str(e)
+                )
+            if content["error_code"] != 0:
+                return create_error_response(content["error_code"], content["text"])
+            choices.append(
+                CompletionResponseChoice(
+                    index=i,
+                    text=content["text"],
+                    logprobs=content.get("logprobs", None),
+                    finish_reason=content.get("finish_reason", "stop"),
+                )
+            )
+            for usage_k, usage_v in content["usage"].items():
```

Comment:
same as the comment in `create_chat_completion`, maybe parse `content["usage"]` directly